### DISCUSSION TITLE
Class Decorators

### DISCUSSION THREAD

**Discussion by hmc-cs-mdrissi** [microsoft/pyright#1898] [Q&A] [answered]
Does pyright support class decorators and if so what's the right way to type annotate them. The code looks something like,

```
def make_cached_object(base_cls):
    class CachedObject(base_cls):
        object_cache = {}

        @classmethod
        def create(cls, *args: object, **kwargs: object):
            cache_key = _make_cache_key(args, kwargs)

            if cache_key in cls.object_cache:
                return cls.object_cache[cache_key]

            obj = CachedObject(*args, **kwargs)
            cls.object_cache[cache_key] = obj
            return obj

        def __init__(self, *args: object, **kwargs: object):
            super(CachedObject, self).__init__(*args, **kwargs)

    return CachedObject
```

Only missing code is make_cache_key and I can include working example if needed, but ideal type is something like base_cls is a class type, return output is CachedObject that's generic over that same class type and create returns a value of that class type. I've played with typevars but have gotten stuck trying to make it work and am unsure if it's possible or not. My type attempt look like this and has multiple errors,

```
T = TypeVar('T')

def make_cached_object(base_cls: T) -> 'Type[CachedObject[T]]':
    class CachedObject(base_cls, Generic[T]):
        object_cache = {}

        @classmethod
        def create(cls, *args: object, **kwargs: object) -> 'CachedObject[T]':
            cache_key = _make_cache_key(args, kwargs)

            if cache_key in cls.object_cache:
                return cls.object_cache[cache_key]

            obj = CachedObject(*args, **kwargs)
            cls.object_cache[cache_key] = obj
            return obj

        def __init__(self, *args: object, **kwargs: object):
            super(CachedObject, self).__init__(*args, **kwargs)

    return CachedObject
```

**Comment by erictraut** [comment, marked as answer]
Pyright doesn't support the use of dynamic arguments in a class declaration. Neither does [mypy](https://github.com/python/mypy/issues/5865). Static type checkers need to understand a class's full inheritance hierarchy, so if you're dynamically generating that hierarchy, the type checker won't be able to reason about the resulting class.

If your intent is to return a class that is effectively the same as the class it is decorating (at least from an interface perspective), then you can annotate it as follows:

```python
def make_cached_object(base_cls: Type[T]) -> Type[T]:
    class CachedObject(base_cls):  # type: ignore
        ....

    return cast(Type[T], CachedObject)
```

**Reply by hmc-cs-mdrissi** [reply]
That works well enough. I mostly wanted to keep the class interface for good IDE suggestions.
